File: packages/observing-api/src/routes/occurrences.ts
/**
 * Occurrence routes - read and write endpoints
 */

import { Router } from "express";
import { Database, getIdentityResolver, TaxonomyClient, GeocodingService } from "observing-shared";
import { enrichOccurrences, enrichIdentifications, enrichComments } from "../enrichment.js";
import { logger } from "../middleware/logging.js";
import { requireAuth } from "../middleware/auth.js";
import { InternalClient } from "../internal-client.js";

export function createOccurrenceRoutes(
  db: Database,
  taxonomy: TaxonomyClient,
  geocoding: GeocodingService,
  internalClient: InternalClient
): Router {
  const router = Router();

  // Get observers for an occurrence
  router.get("/:uri(*)/observers", async (req, res) => {
    try {
      const occurrenceUri = req.params["uri"];
      if (!occurrenceUri) {
        res.status(400).json({ error: "uri is required" });
        return;
      }

      const occurrence = await db.getOccurrence(occurrenceUri);
      if (!occurrence) {
        res.status(404).json({ error: "Occurrence not found" });
        return;
      }

      const observerData = await db.getOccurrenceObservers(occurrenceUri);

      // Enrich with profile info
      const resolver = getIdentityResolver();
      const dids = observerData.map((o) => o.did);
      const profiles = await resolver.getProfiles(dids);

      const observers = observerData.map((o) => {
        const profile = profiles.get(o.did);
        return {
          did: o.did,
          handle: profile?.handle,
          displayName: profile?.displayName,
          avatar: profile?.avatar,
          role: o.role,
          addedAt: o.addedAt.toISOString(),
        };
      });

      res.json({ observers });
    } catch (error) {
      logger.error({ err: error }, "Error fetching observers");
      res.status(500).json({ error: "Internal server error" });
    }
  });

  // Get occurrences nearby
  router.get("/nearby", async (req, res) => {
    try {
      const lat = parseFloat(req.query["lat"] as string);
      const lng = parseFloat(req.query["lng"] as string);
      const radius = parseFloat(req.query["radius"] as string) || 10000;
      const limit = parseInt(req.query["limit"] as string) || 100;
      const offset = parseInt(req.query["offset"] as string) || 0;

      if (isNaN(lat) || isNaN(lng)) {
        res.status(400).json({ error: "lat and lng are required" });
        return;
      }

      if (lat < -90 || lat > 90 || lng < -180 || lng > 180) {
        res.status(400).json({ error: "Invalid coordinates" });
        return;
      }

      const rows = await db.getOccurrencesNearby(lat, lng, radius, limit, offset);
      const occurrences = await enrichOccurrences(db, rows);

      res.json({
        occurrences,
        meta: {
          lat,
          lng,
          radius,
          limit,
          offset,
          count: occurrences.length,
        },
      });
    } catch (error) {
      logger.error({ err: error }, "Error fetching nearby occurrences");
      res.status(500).json({ error: "Internal server error" });
    }
  });

  // Get occurrences feed (chronological)
  router.get("/feed", async (req, res) => {
    try {
      const limit = Math.min(parseInt(req.query["limit"] as string) || 20, 100);
      const cursor = req.query["cursor"] as string | undefined;

      const rows = await db.getOccurrencesFeed(limit, cursor);
      const occurrences = await enrichOccurrences(db, rows);

      const lastRow = rows[rows.length - 1];
      const nextCursor =
        rows.length === limit && lastRow
          ? lastRow.created_at.toISOString()
          : undefined;

      res.json({
        occurrences,
        cursor: nextCursor,
      });
    } catch (error) {
      logger.error({ err: error }, "Error fetching feed");
      res.status(500).json({ error: "Internal server error" });
    }
  });

  // Get occurrences in bounding box
  router.get("/bbox", async (req, res) => {
    try {
      const minLat = parseFloat(req.query["minLat"] as string);
      const minLng = parseFloat(req.query["minLng"] as string);
      const maxLat = parseFloat(req.query["maxLat"] as string);
      const maxLng = parseFloat(req.query["maxLng"] as string);
      const limit = parseInt(req.query["limit"] as string) || 1000;

      if (isNaN(minLat) || isNaN(minLng) || isNaN(maxLat) || isNaN(maxLng)) {
        res.status(400).json({
          error: "minLat, minLng, maxLat, maxLng are required",
        });
        return;
      }

      const rows = await db.getOccurrencesByBoundingBox(
        minLat,
        minLng,
        maxLat,
        maxLng,
        limit
      );

      const occurrences = await enrichOccurrences(db, rows);

      res.json({
        occurrences,
        meta: {
          bounds: { minLat, minLng, maxLat, maxLng },
          count: occurrences.length,
        },
      });
    } catch (error) {
      logger.error({ err: error }, "Error fetching bbox occurrences");
      res.status(500).json({ error: "Internal server error" });
    }
  });

  // Get GeoJSON for map clustering (must be before :uri(*) route)
  router.get("/geojson", async (req, res) => {
    try {
      const minLat = parseFloat(req.query["minLat"] as string);
      const minLng = parseFloat(req.query["minLng"] as string);
      const maxLat = parseFloat(req.query["maxLat"] as string);
      const maxLng = parseFloat(req.query["maxLng"] as string);

      if (isNaN(minLat) || isNaN(minLng) || isNaN(maxLat) || isNaN(maxLng)) {
        res.status(400).json({ error: "Bounding box required" });
        return;
      }

      const rows = await db.getOccurrencesByBoundingBox(
        minLat,
        minLng,
        maxLat,
        maxLng,
        5000
      );

      const features = rows.map((row) => ({
        type: "Feature" as const,
        geometry: {
          type: "Point" as const,
          coordinates: [row.longitude, row.latitude],
        },
        properties: {
          uri: row.uri,
          scientificName: row.scientific_name,
          eventDate: row.event_date.toISOString(),
        },
      }));

      res.json({
        type: "FeatureCollection",
        features,
      });
    } catch (error) {
      logger.error({ err: error }, "Error generating GeoJSON");
      res.status(500).json({ error: "Internal server error" });
    }
  });

  // Create a new occurrence
  router.post("/", requireAuth, async (req, res) => {
    try {
      const sessionDid = req.user!.did;
      const {
        scientificName,
        latitude,
        longitude,
        notes,
        license,
        eventDate,
        images,
        taxonId,
        taxonRank,
        vernacularName,
        kingdom,
        phylum,
        class: taxonomyClass,
        order,
        family,
        genus,
        recordedBy,
      } = req.body;

      if (!latitude || !longitude) {
        res.status(400).json({ error: "latitude and longitude are required" });
        return;
      }

      // Upload images as blobs if provided
      const associatedMedia: Array<{ image: unknown; alt: string }> = [];

      if (images && Array.isArray(images)) {
        for (let i = 0; i < images.length; i++) {
          const img = images[i] as { data: string; mimeType: string };
          if (!img.data || !img.mimeType) continue;

          const blobResult = await internalClient.uploadBlob(sessionDid, img.data, img.mimeType);
          if (blobResult.success && blobResult.blob) {
            associatedMedia.push({
              image: blobResult.blob,
              alt: `Photo ${i + 1}${scientificName ? ` of ${scientificName}` : ""}`,
            });
            logger.info({ mimeType: img.mimeType }, "Uploaded blob via internal RPC");
          } else {
            logger.error({ error: blobResult.error }, "Failed to upload blob");
          }
        }
      }

      // Fetch taxonomy hierarchy from GBIF if scientificName is provided (for identification record)
      let taxon: {
        id?: string | undefined;
        commonName?: string | undefined;
        kingdom?: string | undefined;
        phylum?: string | undefined;
        class?: string | undefined;
        order?: string | undefined;
        family?: string | undefined;
        genus?: string | undefined;
        rank?: string | undefined;
      } | undefined;
      if (scientificName && !taxonId) {
        const validationResult = await taxonomy.validate(scientificName.trim());
        taxon = validationResult.taxon;
      }

      // Reverse geocode to get administrative geography fields
      const geocoded = await geocoding.reverseGeocode(latitude, longitude);

      // Build the occurrence record WITHOUT taxonomy fields
      // Taxonomy data goes into the identification record instead
      const record: Record<string, unknown> = {
        $type: "org.rwell.test.occurrence",
        eventDate: eventDate || new Date().toISOString(),
        location: {
          decimalLatitude: String(latitude),
          decimalLongitude: String(longitude),
          coordinateUncertaintyInMeters: 50,
          geodeticDatum: "WGS84",
          continent: geocoded.continent,
          country: geocoded.country,
          countryCode: geocoded.countryCode,
          stateProvince: geocoded.stateProvince,
          county: geocoded.county,
          municipality: geocoded.municipality,
          locality: geocoded.locality,
          waterBody: geocoded.waterBody,
        },
        notes: notes || undefined,
        license: license || undefined,
        createdAt: new Date().toISOString(),
      };

      if (associatedMedia.length > 0) {
        record["associatedMedia"] = associatedMedia;
      }

      // Add co-observers if provided
      const coObservers: string[] = [];
      if (recordedBy && Array.isArray(recordedBy)) {
        for (const did of recordedBy) {
          if (typeof did === "string" && did !== sessionDid) {
            coObservers.push(did);
          }
        }
        if (coObservers.length > 0) {
          record["recordedBy"] = coObservers;
        }
      }

      // Create the record via internal RPC
      const result = await internalClient.createRecord(
        sessionDid,
        "org.rwell.test.occurrence",
        record
      );

      if (!result.success || !result.uri) {
        res.status(500).json({ error: result.error || "Failed to create record" });
        return;
      }

      logger.info({ uri: result.uri, imageCount: associatedMedia.length }, "Created occurrence via internal RPC");

      // Store exact coordinates in private data table
      await db.saveOccurrencePrivateData(
        result.uri,
        latitude,
        longitude,
        "open"
      );

      // Sync observers table
      await db.syncOccurrenceObservers(result.uri, sessionDid, coObservers);

      // Create identification record if scientificName was provided
      let identificationUri: string | undefined;
      let identificationCid: string | undefined;
      if (scientificName) {
        const identificationRecord = {
          $type: "org.rwell.test.identification",
          subject: {
            uri: result.uri,
            cid: result.cid,
          },
          subjectIndex: 0,
          taxonName: scientificName.trim(),
          taxonRank: taxonRank || taxon?.rank || "species",
          isAgreement: false,
          confidence: "high",
          createdAt: new Date().toISOString(),
          taxonId: taxonId || taxon?.id,
          vernacularName: vernacularName || taxon?.commonName,
          kingdom: kingdom || taxon?.kingdom,
          phylum: phylum || taxon?.phylum,
          class: taxonomyClass || taxon?.class,
          order: order || taxon?.order,
          family: family || taxon?.family,
          genus: genus || taxon?.genus,
        };

        const idResult = await internalClient.createRecord(
          sessionDid,
          "org.rwell.test.identification",
          identificationRecord
        );

        if (idResult.success && idResult.uri) {
          identificationUri = idResult.uri;
          identificationCid = idResult.cid;
          logger.info({ uri: idResult.uri, occurrenceUri: result.uri }, "Created initial identification via internal RPC");
        } else {
          logger.error({ error: idResult.error }, "Failed to create initial identification");
        }
      }

      res.status(201).json({
        success: true,
        uri: result.uri,
        cid: result.cid,
        identificationUri,
        identificationCid,
        message: "Observation posted to AT Protocol network",
      });
    } catch (error) {
      logger.error({ err: error }, "Error creating occurrence");
      res.status(500).json({ error: "Internal server error" });
    }
  });

  // Get single occurrence (must be after specific routes like /geojson)
  router.get("/:uri(*)", async (req, res) => {
    try {
      const uri = req.params["uri"];
      if (!uri) {
        res.status(400).json({ error: "uri is required" });
        return;
      }
      const row = await db.getOccurrence(uri);

      if (!row) {
        res.status(404).json({ error: "Occurrence not found" });
        return;
      }

      const [occurrence] = await enrichOccurrences(db, [row]);
      const identifications = await db.getIdentificationsForOccurrence(uri);
      const comments = await db.getCommentsForOccurrence(uri);

      res.json({
        occurrence,
        identifications: await enrichIdentifications(identifications),
        comments: await enrichComments(comments),
      });
    } catch (error) {
      logger.error({ err: error }, "Error fetching occurrence");
      res.status(500).json({ error: "Internal server error" });
    }
  });

  return router;
}

File: crates/observing-db/src/occurrences.rs
use crate::types::{OccurrenceRow, UpsertOccurrenceParams};

/// Standard SELECT columns for OccurrenceRow in QueryBuilder (runtime) queries.
/// Does not include the SELECT keyword or FROM clause.
/// Use with `concat!` for compile-time string composition:
/// ```ignore
/// concat!("SELECT ", occurrence_columns!(), " FROM occurrences WHERE TRUE")
/// ```
#[macro_export]
macro_rules! occurrence_columns {
    () => {
        r#"
    uri, cid, did, scientific_name, event_date,
    ST_Y(location::geometry) as latitude,
    ST_X(location::geometry) as longitude,
    coordinate_uncertainty_meters,
    continent, country, country_code, state_province, county, municipality, locality, water_body,
    verbatim_locality, occurrence_remarks,
    associated_media, recorded_by,
    taxon_id, taxon_rank, vernacular_name, kingdom, phylum, class, "order", family, genus,
    created_at,
    NULL::float8 as distance_meters,
    NULL::text as source,
    NULL::text as observer_role
"#
    };
}

/// Upsert an occurrence record
pub async fn upsert(
    executor: impl sqlx::PgExecutor<'_>,
    p: &UpsertOccurrenceParams,
) -> Result<(), sqlx::Error> {
    sqlx::query!(
        r#"
        INSERT INTO occurrences (
            uri, cid, did, scientific_name, event_date, location,
            coordinate_uncertainty_meters,
            continent, country, country_code, state_province, county, municipality, locality, water_body,
            verbatim_locality, occurrence_remarks, associated_media, recorded_by,
            taxon_id, taxon_rank, vernacular_name, kingdom, phylum, class, "order", family, genus,
            created_at
        ) VALUES (
            $1, $2, $3, $4, $5,
            ST_SetSRID(ST_MakePoint($6, $7), 4326)::geography,
            $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20,
            $21, $22, $23, $24, $25, $26, $27, $28, $29, $30
        )
        ON CONFLICT (uri) DO UPDATE SET
            cid = $2,
            scientific_name = $4,
            event_date = $5,
            location = ST_SetSRID(ST_MakePoint($6, $7), 4326)::geography,
            coordinate_uncertainty_meters = $8,
            continent = COALESCE($9, occurrences.continent),
            country = COALESCE($10, occurrences.country),
            country_code = COALESCE($11, occurrences.country_code),
            state_province = COALESCE($12, occurrences.state_province),
            county = COALESCE($13, occurrences.county),
            municipality = COALESCE($14, occurrences.municipality),
            locality = COALESCE($15, occurrences.locality),
            water_body = COALESCE($16, occurrences.water_body),
            verbatim_locality = $17,
            occurrence_remarks = $18,
            associated_media = COALESCE($19, occurrences.associated_media),
            recorded_by = COALESCE($20, occurrences.recorded_by),
            taxon_id = COALESCE($21, occurrences.taxon_id),
            taxon_rank = COALESCE($22, occurrences.taxon_rank),
            vernacular_name = COALESCE($23, occurrences.vernacular_name),
            kingdom = COALESCE($24, occurrences.kingdom),
            phylum = COALESCE($25, occurrences.phylum),
            class = COALESCE($26, occurrences.class),
            "order" = COALESCE($27, occurrences."order"),
            family = COALESCE($28, occurrences.family),
            genus = COALESCE($29, occurrences.genus),
            indexed_at = NOW()
        "#,
        p.uri,
        p.cid,
        p.did,
        p.scientific_name as _,
        p.event_date,
        p.longitude,
        p.latitude,
        p.coordinate_uncertainty_meters as _,
        p.continent as _,
        p.country as _,
        p.country_code as _,
        p.state_province as _,
        p.county as _,
        p.municipality as _,
        p.locality as _,
        p.water_body as _,
        p.verbatim_locality as _,
        p.occurrence_remarks as _,
        p.associated_media as _,
        p.recorded_by as _,
        p.taxon_id as _,
        p.taxon_rank as _,
        p.vernacular_name as _,
        p.kingdom as _,
        p.phylum as _,
        p.class as _,
        p.order as _,
        p.family as _,
        p.genus as _,
        p.created_at,
    )
    .execute(executor)
    .await?;
    Ok(())
}

/// Delete an occurrence
pub async fn delete(executor: impl sqlx::PgExecutor<'_>, uri: &str) -> Result<(), sqlx::Error> {
    sqlx::query!("DELETE FROM occurrences WHERE uri = $1", uri)
        .execute(executor)
        .await?;
    Ok(())
}

/// Get a single occurrence by URI
pub async fn get(
    executor: impl sqlx::PgExecutor<'_>,
    uri: &str,
) -> Result<Option<OccurrenceRow>, sqlx::Error> {
    sqlx::query_as!(
        OccurrenceRow,
        r#"
        SELECT
            uri, cid, did, scientific_name, event_date,
            ST_Y(location::geometry) as "latitude!",
            ST_X(location::geometry) as "longitude!",
            coordinate_uncertainty_meters,
            continent, country, country_code, state_province, county, municipality, locality, water_body,
            verbatim_locality, occurrence_remarks,
            associated_media, recorded_by,
            taxon_id, taxon_rank, vernacular_name, kingdom, phylum, class, "order" as order_, family, genus,
            created_at,
            NULL::float8 as distance_meters,
            NULL::text as source,
            NULL::text as observer_role
        FROM occurrences
        WHERE uri = $1
        "#,
        uri,
    )
    .fetch_optional(executor)
    .await
}

/// Get occurrences nearby a point
pub async fn get_nearby(
    executor: impl sqlx::PgExecutor<'_>,
    lat: f64,
    lng: f64,
    radius_meters: f64,
    limit: i64,
    offset: i64,
    hidden_dids: &[String],
) -> Result<Vec<OccurrenceRow>, sqlx::Error> {
    sqlx::query_as!(
        OccurrenceRow,
        r#"
        SELECT
            uri, cid, did, scientific_name, event_date,
            ST_Y(location::geometry) as "latitude!",
            ST_X(location::geometry) as "longitude!",
            coordinate_uncertainty_meters,
            continent, country, country_code, state_province, county, municipality, locality, water_body,
            verbatim_locality, occurrence_remarks,
            associated_media, recorded_by,
            taxon_id, taxon_rank, vernacular_name, kingdom, phylum, class, "order" as order_, family, genus,
            created_at,
            ST_Distance(location, ST_SetSRID(ST_MakePoint($2, $1), 4326)::geography) as distance_meters,
            NULL::text as source,
            NULL::text as observer_role
        FROM occurrences
        WHERE ST_DWithin(
            location,
            ST_SetSRID(ST_MakePoint($2, $1), 4326)::geography,
            $3
        )
        AND did != ALL($6)
        ORDER BY distance_meters
        LIMIT $4 OFFSET $5
        "#,
        lat,
        lng,
        radius_meters,
        limit,
        offset,
        hidden_dids,
    )
    .fetch_all(executor)
    .await
}

/// Get occurrences within a bounding box
pub async fn get_by_bounding_box(
    executor: impl sqlx::PgExecutor<'_>,
    min_lat: f64,
    min_lng: f64,
    max_lat: f64,
    max_lng: f64,
    limit: i64,
    hidden_dids: &[String],
) -> Result<Vec<OccurrenceRow>, sqlx::Error> {
    sqlx::query_as!(
        OccurrenceRow,
        r#"
        SELECT
            uri, cid, did, scientific_name, event_date,
            ST_Y(location::geometry) as "latitude!",
            ST_X(location::geometry) as "longitude!",
            coordinate_uncertainty_meters,
            continent, country, country_code, state_province, county, municipality, locality, water_body,
            verbatim_locality, occurrence_remarks,
            associated_media, recorded_by,
            taxon_id, taxon_rank, vernacular_name, kingdom, phylum, class, "order" as order_, family, genus,
            created_at,
            NULL::float8 as distance_meters,
            NULL::text as source,
            NULL::text as observer_role
        FROM occurrences
        WHERE location && ST_MakeEnvelope($1, $2, $3, $4, 4326)::geography
        AND did != ALL($6)
        LIMIT $5
        "#,
        min_lng,
        min_lat,
        max_lng,
        max_lat,
        limit,
        hidden_dids,
    )
    .fetch_all(executor)
    .await
}

/// Get occurrences feed (chronological, cursor-based)
pub async fn get_feed(
    executor: impl sqlx::PgExecutor<'_>,
    limit: i64,
    cursor: Option<&str>,
    hidden_dids: &[String],
) -> Result<Vec<OccurrenceRow>, sqlx::Error> {
    if let Some(cursor) = cursor {
        sqlx::query_as!(
            OccurrenceRow,
            r#"
            SELECT
                uri, cid, did, scientific_name, event_date,
                ST_Y(location::geometry) as "latitude!",
                ST_X(location::geometry) as "longitude!",
                coordinate_uncertainty_meters,
                continent, country, country_code, state_province, county, municipality, locality, water_body,
                verbatim_locality, occurrence_remarks,
                associated_media, recorded_by,
                taxon_id, taxon_rank, vernacular_name, kingdom, phylum, class, "order" as order_, family, genus,
                created_at,
                NULL::float8 as distance_meters,
                NULL::text as source,
                NULL::text as observer_role
            FROM occurrences
            WHERE created_at < ($2::text)::timestamptz
            AND did != ALL($3)
            ORDER BY created_at DESC
            LIMIT $1
            "#,
            limit,
            cursor,
            hidden_dids,
        )
        .fetch_all(executor)
        .await
    } else {
        sqlx::query_as!(
            OccurrenceRow,
            r#"
            SELECT
                uri, cid, did, scientific_name, event_date,
                ST_Y(location::geometry) as "latitude!",
                ST_X(location::geometry) as "longitude!",
                coordinate_uncertainty_meters,
                continent, country, country_code, state_province, county, municipality, locality, water_body,
                verbatim_locality, occurrence_remarks,
                associated_media, recorded_by,
                taxon_id, taxon_rank, vernacular_name, kingdom, phylum, class, "order" as order_, family, genus,
                created_at,
                NULL::float8 as distance_meters,
                NULL::text as source,
                NULL::text as observer_role
            FROM occurrences
            WHERE did != ALL($2)
            ORDER BY created_at DESC
            LIMIT $1
            "#,
            limit,
            hidden_dids,
        )
        .fetch_all(executor)
        .await
    }
}
